File: 943-sum-of-subarray-minimums/sum-of-subarray-minimums.py
from collections import deque
class Solution:
    def sumSubarrayMins(self, arr: List[int]) -> int:
        mod = (10 ** 9 + 7)
        stack = deque()
        n = len(arr)
        total = 0
        prev_smaller = [-1] * n
        
        for i in range(0, n):
            elem = arr[i]
            while (stack and arr[stack[-1]] >= elem):
                stack.pop()
            prev_smaller[i] = stack[-1] if stack else -1
            stack.append(i)
        stack.clear()
        next_smaller = [n] * n

        for i in range(n-1, -1, -1):
            elem = arr[i]
            while (stack and arr[stack[-1]] > elem):
                stack.pop()
            next_smaller[i] = stack[-1] if stack else n
            stack.append(i)

        # print(prev_smaller, next_smaller, sep="\n")
        for i in range(0, n):
            nse = next_smaller[i]
            pse = prev_smaller[i]

            possiblities = (nse - i) * (i - pse)
            total += (possiblities * arr[i])
            total = total % mod
        
        return total

            

File: 943-sum-of-subarray-minimums/sum-of-subarray-minimums.java
class Solution {
    public int[] getNSE(int[] arr, int n) {
        int[] nse = new int[n];
        Deque<Integer> dq = new ArrayDeque<Integer>();

        for(int i = n-1; i>=0; i--) {
            while (!dq.isEmpty() && arr[dq.peekLast()] >= arr[i]) {
                dq.removeLast();
            }
            if (dq.isEmpty()) {
                nse[i] = n;
            }
            else {
                nse[i] = dq.peekLast();
            }
            dq.add(i);
        }
        return nse;
    }

    public int[] getPSE(int[] arr, int n) {
        int[] pse = new int[n];
        Deque<Integer> dq = new ArrayDeque<Integer>();

        for(int i = 0; i<n; i++) {
            while (!dq.isEmpty() && arr[dq.peekLast()] > arr[i]) {
                dq.removeLast();
            }
            if (dq.isEmpty()) {
                pse[i] = -1;
            }
            else {
                pse[i] = dq.peekLast();
            }
            dq.add(i);
        }
        return pse;
    }
    public int sumSubarrayMins(int[] arr) {
        int n = arr.length;
        int[] nse = getNSE(arr, n);
        int[] pse = getPSE(arr, n);
        long res = 0;
        final long MOD = 1_000_000_007;
        for (int i = 0; i<n; i++) {
            long left = nse[i]-i;
            long right = i-pse[i];
            long numElems = (right*left)%MOD;
            long contrib = (numElems*arr[i])%MOD;
            res = (long)((res + contrib)%MOD);
        }
        return (int)res;

    }
}